package.json

```diff
@@ -0,0 +1,23 @@
+{
+    "name": "jupyterlab-jupytext-extensions",
+    "version": "1.3.9",
+    "private": true,
+    "files": [],
+    "workspaces": [
+        "packages/*"
+    ],
```

Comment:
Wouldn't that structure be for `jupyterlab-jupytext` sub-package? The other reference packages are single package right?

Other than that it is best to have a `src-layout`, particularly if you have a multi-ptoject directory structure like this. Otherwise you risk having unpackaged modules that seem like it would work, but they only work because it is being run in the git repo working directory.

Reply:
If we split the package into `jupytext` and `jupytext_config` modules as you suggested, yes, I agree that we can move them into a top level `src/` and explicitly specify modules in `pyproject.toml` so that the risk of _unpackaged_ modules will be lowered. However, I do not see a strong reason to change structure of `packages/` folder as we do not pick up any modules in `packages/` folder from top level repo and _vice-versa_. Moreover, each package _should_ distribute only _one_ module.

In any ways, we need to _explicitly_ specify packages in `publish.yaml` workflow to publish them and even if we use `jupyter-releaser`, we need to set `python-packages` in `jupyter-releaser` config to build and publish them. So, the risk of unpackaged modules will be lower. What do you think?

Reply:
The risk is with tests, where we think everything would work well, but when it is actually packaged...

About `packages`, it should still be there, but in the sub-project instead of the top-project location. Actually `jupyter-config` as well? `demo` and `binder` I am not sure, need to see how it is invoked. The idea is that the top-level package which does not have the direct jupyter packaging should be as vanilla in its structure as possible.

Reply:
I do not understand what do you exactly mean. In your original comment, this structure

```
├── jupyterlab-jupytext
│   ├── packages
│   └── src
│       └── jupyterlab_jupytext
```

is unclear to me. What do you put in `packages/` and `src/jupyterlab_jupytext` in this structure?

Reply:
Sry, I've truncated it because I was writing it manually. `packages` contains all of the contents that are currently there. `src/jupyterlab_jupytext` has the python module. An equivalent `src-layput` for multi-language projects is to have `pytgon/jupyterlab_jupytext`.

Reply:
No problem. For instance, it is the current state of this PR:

```
├── binder
├── demo
├── docs
├── jupyter-config
├── jupytext
├── packages
│   └── labextension
│       ├── jupyterlab_jupytext
│       └── src
└── tests
```
So, the `labextension` folder has the standalone python package that installs JupyterLab extension. If we develop more extensions in the future, we will put them under `packages/` folder as separate python packages.

> packages contains all of the contents that are currently there. src/jupyterlab_jupytext has the python module. 

So, `packages` has _already_ the python module/package, so there is nothing to put in `src/jupyterlab_jupytext`. The way I am looking at the structure is as follows:

```
.
├── binder
├── demo
├── docs
├── packages
│   └── jupyterlab_jupytext
|       |── jupyter-config
│       ├── package.json
│       ├── pyproject.toml
│       └── src
├── pyproject.toml
└── src
    ├── jupytext
    └── jupytext_config
``` 

Top level modules go into `src` and Jupyter related packages go into `packages`. Do you agree with it?

I understand the idea of keeping it as vanilla as possible, but we need to support jupyter ecosystem as well within the repo. And this is sort of standard organisation that they use in most of their extensions.

Reply:
Sorry I am afraid I am the least knowledgeable person on this thread. I am fine with the current directory structure and I agree that the `packages/jupyterlab_jupytext` folder has to follow the established Jupyter practice. Other than that I let it up to you to decide what works best - thanks!

Reply:
Ok I've looked around a bit and I understand what you mean with `packages`, but I think there's a bit of misunderstanding about sub-projects.

@mahendrapaipuri in your structure you have:
- each `package` is its own sub-project which would have its own PyPI entry
- the main jupyterlab extension would still be in the top-level project

What I was thinking is to move the whole jupyterlab extension in the `jupyterlab` folder (renamed the folder to make clear this is not the `jupyterlab/packages/jupyterlab-jupytext`. The latter structure is still there), with `packages.json` and so on in that folder. Some upstream actions would require a `path` input to run in the appropriate folder, but I can add those when we have a set of tools to use.

The `jupyterlab/packages` would still be extendable if we want to add more jupyter extensions with different `package.json`, but they would use the `jupyterlab/pyproject.toml`. I am not sure if we should further split so that each `packages` to be their own `pyproject.toml`. We can do so if we find some use-cases, but let's try it out like this first.

The main scope of this split is to have the toolings separated in each one, where top-level has vanilla python tools, packaging and testing, while the sub-project requires jupyter specific ones.

Reply:
Yes exactly, `jupytext` in `pypi` or `conda` will always have the extension.

Actually @mahendrapaipuri you can very well remove the environment variable, these days the Jupyter Lab extension is an important part of Jupytext so I would be fine with including it even in the development builds.

I do have a preference for the one package approach, is it fine with you too? Do you see a downside to it?

Reply:
Please check updated [PR desc](https://github.com/mwouts/jupytext/pull/1134#issue-1941666589). Essentially, we can have a single package that distributes multiple python modules split based on the functionality they provide. If you prefer to have extension build by default in dev env, we can do it as well. We use the same env var to _skip_ building the labextension if developer wishes

Reply:
@mwouts The only downside I see is that when a developer wants to _use_ extension while _developing_ core `jupytext`, (s)he _must_ build the extension in their local dev env. In a separate package approach, they can simply install a _python_ package that installs the extension without having to build it.

But I guess we can live with building extension when needed!!